functions/Invoke-DbaLogShippingRecovery.ps1

```diff
@@ -0,0 +1,307 @@
+	)
+
+    BEGIN
+    {
+        Write-Output "Attempting to connect to Sql Server.."
+		$server = Connect-SqlServer -SqlServer $SqlServer -SqlCredential $SqlCredential
+    }
+
+    PROCESS
+    {
+        Write-Output "Started Log Shipping Recovery."
+
+        # Query for retrieving the log shipping information
+        $query = "
+        SELECT  lss.primary_server ,
+                lss.primary_database ,
+                lsd.secondary_database ,
+                lss.backup_source_directory ,
+                lss.backup_destination_directory ,
+                lss.last_copied_file ,
+                lss.last_copied_date ,
+                lsd.last_restored_file,
+		        sj1.name AS 'copyjob',
+		        sj2.name AS 'restorejob'
+        FROM    msdb.dbo.log_shipping_secondary AS lss
+                INNER JOIN msdb.dbo.log_shipping_secondary_databases AS lsd ON lsd.secondary_id = lss.secondary_id
+		        INNER JOIN msdb.dbo.sysjobs AS sj1 ON sj1.job_id = lss.copy_job_id
+		        INNER JOIN msdb.dbo.sysjobs AS sj2 ON sj2.job_id = lss.restore_job_id "
+        
+        # Check if they commandlet is eecuted from a pipeline
+		if ($PSCmdlet.MyInvocation.ExpectingInput) 
+		{
+			$db = $_
+            
+            $query += "WHERE lsd.secondary_database = '$db'"
+		}
+		else
+		{
+            # If just one database is set in the database parameter
+			if($Database -ne $null)
+            {
+			    $db = $Database[0]
+
+                $query += "WHERE lsd.secondary_database = '$db'"
+            }
+		}
+
+        # Retrieve the log shipping information from the secondary instance
+        try{
+            Write-Verbose "Retrieving log shipping information from the secondary instance"
+            $logshipping_details = Invoke-Sqlcmd2 -ServerInstance $SqlServer -Database 'msdb' -Query $query
+        }
+        catch
+        {
+            Write-Error $_.Exception.Message
+            Break
+        }
+        
+        # Check if there are any databases to recover
+        if($logshipping_details -ne $null)
+        {
+
+            # Loop through each of the log shipped databases
+            foreach($ls in $logshipping_details)
+            {
+                # Check if the database is in the right state
+                if($server.Databases[$ls.secondary_database].Status -in ('Normal, Standby', 'Standby', 'Restoring'))
+                {
+
+                    Write-Output ("Started Recovery for '" + $ls.secondary_database + "'")
+
+                    # Check if the backup source directory can be reached
+                    if(Test-Path $ls.backup_source_directory)
+                    {
+                        # Get the last file from the backup source directory
+                        $latestfile = Get-ChildItem -Path $ls.backup_source_directory -filter ("*" + $ls.primary_database + "*") | where {($_.Extension -eq '.trn') } | Sort-Object LastWriteTime -Descending | Select-Object -First 1
```

Comment:
Avoid using Aliases, such as "where", instead of the full cmdlet "Where-Object".
If you use PowerShell Studio as an editor, you can expand aliases with the "Tab" key, dunno about the other editors.